src/tests/tpch/sb_tpch.c

```diff
@@ -0,0 +1,425 @@
+typedef struct tpch_s {
+    unsigned int size;
+    char *root_path;
+    char *query_path;
+    db_driver_t *db_driver;
+    char **sql_queries;
+} tpch_t;
+
+static tpch_t tpch = {};
+
+static sb_test_t tpch_test =
+{
+  .sname = "tpch",
+  .lname = "TPC-H performance test",
+  .ops = {
+    .init = tpch_init,
+    .print_mode = tpch_print_mode,
+    .next_event = tpch_next_event,
+    .execute_event = tpch_execute_event,
+    .report_intermediate = tpch_report_intermediate,
+    .report_cumulative = tpch_report_cumulative,
+    .done = tpch_done
+  },
+  .builtin_cmds = {
+    .prepare = tpch_prepare
+  },
+  .args = tpch_args
+};
+
+static void tpch_report_json(const double seconds, sb_stat_t *stat)
+{
+    printf("[\n"
+           "\t{\n"
+           "\t\t\"time\": %d,\n"
+           "\t\t\"threads\": %d,\n"
+           "\t\t\"tps\": %4.2f,\n"
+           "\t\t\"qps\": {\n"
+           "\t\t\t\"total\": %4.2f,\n"
+           "\t\t\t\"reads\": %4.2f,\n"
+           "\t\t\t\"writes\": %4.2f,\n"
+           "\t\t\t\"other\": %4.2f,\n"
+           "\t\t\"latency\": %4.2f,\n"
+           "\t\t\"errors\": %4.2f,\n"
+           "\t\t\"reconnects\": %4.2f\n"
+           "\t}\n"
+           "]\n",
+           (int)seconds,
+           0,
+           stat->events / seconds,
+           (stat->reads + stat->writes + stat->other) / seconds,
+           stat->reads / seconds,
+           stat->writes / seconds,
+           stat->other / seconds,
+           SEC2MS(stat->latency_pct),
+           stat->errors / seconds,
+           stat->reconnects / seconds);
+}
+
+static int get_tpch_args(void)
+{
+    int size = sb_get_value_int("data-size");
+    char *root_path = sb_get_value_string("root-path");
+
+    if (size <= 0) {
+        log_text(LOG_FATAL, "Invalid value of data-size: %d.", size);
+        return 1;
+    }
+    if (root_path == NULL) {
+        log_text(LOG_FATAL, "Invalid value of root-path, got NULL.");
+        return 1;
+    }
+
+    tpch.size = (unsigned int)size;
+    tpch.root_path = root_path;
+    return 0;
+}
+
+static char *add_path_to_root(char *add)
+{
+    int size = strlen(tpch.root_path) + strlen(add) + 2;
+    char *path = malloc(size);
+
+    if (path == NULL) {
+        return NULL;
+    }
+    strcat(path, tpch.root_path);
+    if (strlen(tpch.root_path) > 0 && tpch.root_path[strlen(tpch.root_path) - 1] != '/') {
+        strcat(path, "/");
+    }
+    strcat(path, add);
+    return path;
+}
+
+static char** get_script_arguments(const char *path)
+{
+    const int nb_args = 4;
+    char args_str[nb_args][32] = {
+            "--size",
+            "",
+            "--mysql-params",
+            "tpch"
+    };
+    char **args = malloc(sizeof(char *) * (nb_args + 2));
+    if (args == NULL)
+        return NULL;
+
+    sprintf(args_str[1], "%d", tpch.size);
+
+    args[0] = malloc(strlen(path));
+    if (args[0] == NULL) {
+        free(args);
+        return NULL;
+    }
+
+    strcpy(args[0], path);
```

Comment:
Can this logic be simplified using `strdup`? Since it does an alloc and then a `strcpy`?

Reply:
Correct, I applied this change in d7bfd22. Much easier this way